subsys/bluetooth/gatt_dm.c

```diff
@@ -336,13 +371,15 @@ static uint8_t discovery_process_service(struct bt_gatt_dm *dm,
 	dm->discover_params.type         = BT_GATT_DISCOVER_ATTRIBUTE;
 	dm->discover_params.start_handle = cur_attr->handle + 1;
 	LOG_DBG("Starting descriptors discovery");
-	err = bt_gatt_discover(dm->conn, &(dm->discover_params));
 
-	if (err) {
-		LOG_ERR("Descriptor discover failed, error: %d.", err);
-		discovery_complete_error(dm, -ENOMEM);
-		return BT_GATT_ITER_STOP;
-	}
+	gatt_work_data.dm = dm;
+	k_work_init(&gatt_work_data.work, gatt_discover_work);
```

Comment:
It is dangerous to initialize `k_work` in the callback. If the work has already been scheduled, but the callback is called one more time, it may result to undefined behavior because `k_work_init` doesn't check if the work was scheduled or not. Better to move the `k_work_init` call to `bt_gatt_dm_start` I think.